knowledge_storm/dataclass.py

```diff
@@ -0,0 +1,831 @@
+        if information_index not in self.content:
+            self.need_regenerate_synthesize_output = True
+            self.content.add(information_index)
+
+    def get_all_descendents(self) -> List["KnowledgeNode"]:
+        """
+        Get a list of all descendant nodes.
+
+        Returns:
+            List[KnowledgeNode]: A list of all descendant nodes.
+        """
+        descendents = []
+
+        def collect_descendents(node):
+            for child in node.children:
+                descendents.append(child)
+                collect_descendents(child)
+
+        collect_descendents(self)
+        return descendents
+
+    def get_all_predecessors(self) -> List["KnowledgeNode"]:
+        """
+        Get a list of all predecessor nodes (from current node to root).
+
+        Returns:
+            List[KnowledgeNode]: A list of all predecessor nodes.
+        """
+        predecessors = []
+        current_node = self.parent
+        while current_node is not None:
+            predecessors.append(current_node)
+            current_node = current_node.parent
+        return predecessors
+
+    def to_dict(self):
+        """
+        Converts the KnowledgeNode instance to a dictionary representation.
+
+        Returns:
+            dict: The dictionary representation of the KnowledgeNode.
+        """
+        return {
+            "name": self.name,
+            "content": list(self.content),
+            "children": [child.to_dict() for child in self.children],
+            "parent": self.parent.name if self.parent else None,
+            "synthesize_output": self.synthesize_output,
+            "need_regenerate_synthesize_output": self.need_regenerate_synthesize_output,
+        }
+
+    @classmethod
+    def from_dict(cls, data):
+        """
+        Constructs a KnowledgeNode instance from a dictionary representation.
+
+        Args:
+            data (dict): The dictionary representation of the KnowledgeNode.
+
+        Returns:
+            KnowledgeNode: The constructed KnowledgeNode instance.
+        """
+
+        def helper(cls, data, parent_node=None):
+            if parent_node is not None:
+                assert data["parent"] is not None and data["parent"] == parent_node.name
+            node = cls(
+                name=data["name"],
+                content=data["content"],
+                parent=parent_node,
+                children=None,
+                synthesize_output=data.get("synthesize_output", None),
+                need_regenerate_synthesize_output=data.get(
+                    "need_regenerate_synthesize_output", True
+                ),
+            )
+            for child_data in data["children"]:
+                child_node = helper(cls, child_data, parent_node=node)
+                node.children.append(child_node)
+            return node
+
+        return helper(cls, data)
+
+
+class KnowledgeBase:
+    """
+    Class representing the knowledge base.
```

Comment:
Make this a bit more informative. I don't think most of people understand what we mean by knowledge base here.